### PR TITLE
Ticket 463 -- updated documentation && added SimpleZioExample code into the "example directory"

### LINKED ISSUE #463: DOCUMENTATION - add ZIO Metrics doc to "mesmer/docs" Directory
After the completion of Tickets  #421 / #422 / #423 / #424... the purpose of this Ticket is to:

- created documentation regarding ZIO Metrics (similar to https://github.com/ScalaConsultants/mesmer/blob/main/docs/supported-metrics.md#akka-core), and add this new documentation to this page: https://github.com/ScalaConsultants/mesmer/blob/main/docs/supported-metrics.md

@lgajowy, I may need to sync up with you, in order to explicitly define which metrics should be added to this "Mesmer Supported Metrics" page.

### PR DESCRIPTION
**Pull request by ronaldSC-2022**
Closes #463 (DOCUMENTATION - add ZIO Metrics doc to "mesmer/docs" Directory)
___

## Summary / Purpose of this PR:

1)
Add / update Mesmer documentation for:

- How to collect JVM Metrics from ZIO 2.0.0 (for Mesmer)
- updated **intro.md** file
- add ZIO Metrics (both Default-Jvm-Metrics and others) into the **supported-metrics.md** file

2)
In the "example" directory, I added the **SimpleZioExample** logic (copied from Lukasz' ZIO repo: https://github.com/lgajowy/forkenstein)

- The reason I did this is because I was unable to run Mesmer with Forkenstein locally... Also, I thought it might be nice to have this ZIO Example (with the Akka / Akka Stream Examples).

3)
Added logic in order to run: **sbt "project example" runZioExampleWithOtelAgent**

- For "runZioExampleWithOtelAgent", I followed the logic of both **runExampleWithOtelAgent** and **runStreamExampleWithOtelAgent**.
- After running runZioExampleWithOtelAgent locally, I was able to collect the following ZIO Metrics (see screenshots below):

**_(Screenshot 1)_**
runZioExampleWithOtelAgent Screenshot - Part 1

**_(Screenshot 2)_**
runZioExampleWithOtelAgent Screenshot - Part 2

**_(Screenshot 3)_**
runZioExampleWithOtelAgent Screenshot - Part 3

If there are any questions, or if something looks wrong, then please let me know. 🙂

Sincerely,
Ronald W.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -150,7 +150,8 @@ lazy val example = (project in file("example"))
       akkaTestkit.map(_ % "test") ++
       akkaPersistance ++
       logback ++
-      exampleDependencies
+      exampleDependencies ++
```
Comment (lgajowy) [inline build.sbt:153]: Not a critical one: 

I think that we should either get rid of "exampleDependencies" and just paste the dependencies here or add "zio" to example dependencies. I think the former feels nicer to me

File: build.sbt [hunk 2]
```diff
@@ -243,3 +245,21 @@ def runStreamExampleWithOtelAgent = Command.command("runStreamExampleWithOtelAge
...
   s
 }
+
+def runZioExampleWithOtelAgent = Command.command("runZioExampleWithOtelAgent") { state =>
+  val extracted = Project extract state
+  val newState = extracted.appendWithSession(
+    Seq(
+      run / javaOptions ++= Seq(
+        s"-javaagent:$projectRootDir/opentelemetry-javaagent-$OpentelemetryLatestVersion.jar",
+        s"-Dotel.service.name=mesmer-stream-example",
```
Comment (lgajowy) [inline build.sbt:255]: incorrect name

File: docs/getting-started.md [hunk 3]
```diff
@@ -64,3 +64,14 @@ Both parts need to be included in the application for Mesmer to work.
    ```sh
    curl -i http://localhost:9464
    ```
+
+## How to collect JVM Metrics from ZIO (for Mesmer):
+
+In your ZIO Application, you will need to add the following lines of code:
```
Comment (lgajowy) [inline docs/getting-started.md:70]:
```suggestion
In your ZIO Application, you will need to add the following layers:
```

File: docs/intro.md [hunk 4]
```diff
@@ -11,3 +11,30 @@ Auto-instrumentation means it will automatically detect the libraries you use an
...
+# Links to get started:
+
+## ByteBuddy
+This is the tool that we use to instrument java bytecode. It provides a declarative API and allows to register transformation on loaded classes.
+main page tutorial, provides a good overview, but we don't use many features from here (and not all features that we use are there described) - https://bytebuddy.net/#/tutorial
+https://www.youtube.com/watch?v=OF3YFGZcQkg - talk by the creator. Provides good introduction to Advice - tool that we heavily use in mesmer
+
+
+## OpenTelemetry
+Initiative that goal is to create spec for observable systems and provide tools to implement it for different languages.
```
Comment (lgajowy) [inline docs/intro.md:24]:
```suggestion
Initiative that's goal is to create a spec for observable systems and provide tools to implement it for different languages.
```

File: docs/intro.md [hunk 5]
```diff
@@ -11,3 +11,30 @@ Auto-instrumentation means it will automatically detect the libraries you use an
 OpenTelemetry is a set of tools, APIs and SDKs that provide standardized approach to telemetry including: instrumenting, generating, collecting and exporting telemetry data.
 
 Mesmer is designed as an extension for the OpenTelemetry Java agent. This way you will benefit from both the general Java APIs metrics that are still useful in Scala (like JDBC or Netty) and the specific Scala instrumentation provided by Mesmer (like Akka or ZIO).
+
+# Links to get started:
+
+## ByteBuddy
```
Comment (lgajowy) [inline docs/intro.md:17]: Not super important: all the links below could use `[Link makrdown support](HTTP://foo.bar)`

File: docs/supported-metrics.md [hunk 6]
```diff
@@ -54,3 +54,48 @@ In Mesmer we support 3 types of metrics:
...
 | Operator processing time     | counter |
+
+## ZIO (both Default-JVM-Metrics and others)
+
+| Metric                                | Type    |
+|---------------------------------------|---------|
+| Executor Capacity                     | gauge   |
+| Executor Concurrency                  | gauge   |
+| Executor Dequeued Count               |         |
+| Executor Enqueued Count               |         |
```
Comment (lgajowy) [inline docs/supported-metrics.md:65]: These are "Counters" right?

File: docs/supported-metrics.md [hunk 7]
```diff
@@ -54,3 +54,48 @@ In Mesmer we support 3 types of metrics:
...
+| Executor Size                         | gauge   |
+| Executor Worker Count                 |         |
+| JVM Buffer Pool Capacity Bytes        | gauge   |
+| JVM Buffer Pool Used Buffers          | gauge   |
+| JVM Buffer Pool Used Bytes            | gauge   |
+| JVM Classes Loaded                    | gauge   |
+| JVM Classes Loaded Total              | gauge   |
+| JVM Classes Unloaded Total            | gauge   |
+| JVM GC Collection Seconds Count       |         |
+| JVM GC Collection Seconds Sum         |         |
```
Comment (lgajowy) [inline docs/supported-metrics.md:75]: Histogram?

Comment (lgajowy) [reply docs/supported-metrics.md:75]: if that's a histogram then there's _count, _sum and _buckets metric in prometheus.

File: docs/supported-metrics.md [hunk 8]
```diff
@@ -54,3 +54,48 @@ In Mesmer we support 3 types of metrics:
...
+## ZIO (both Default-JVM-Metrics and others)
+
+| Metric                                | Type    |
+|---------------------------------------|---------|
+| Executor Capacity                     | gauge   |
+| Executor Concurrency                  | gauge   |
+| Executor Dequeued Count               |         |
+| Executor Enqueued Count               |         |
+| Executor Size                         | gauge   |
+| Executor Worker Count                 |         |
```
Comment (lgajowy) [inline docs/supported-metrics.md:67]: counter?

File: example/src/main/scala/example/SimpleZioExample.scala [hunk 9]
```diff
@@ -0,0 +1,17 @@
+package example
+
+import zio._
+import zio.metrics.jvm.DefaultJvmMetrics
+
+//NOTE: code originally written by Łukasz Gajowy
```
Comment (lgajowy) [inline example/src/main/scala/example/SimpleZioExample.scala:6]:
```suggestion
```

I appreciate it but no need to add this. :)

**Review by lgajowy — COMMENTED** [review]
Awesome job. Only some minor comments. Please take a look. Thanks! :)

File: docs/supported-metrics.md (hunk 6 above)
Comment (ronaldSC-2022) [reply docs/supported-metrics.md:65]: You are right, these are Counters.

@lgajowy, I will update this MD File for these 2 Metrics.

File: docs/supported-metrics.md [hunk 10]
```diff
@@ -54,3 +54,48 @@ In Mesmer we support 3 types of metrics:
...
+| JVM Memory Pool Bytes Init            | gauge   |
+| JVM Memory Pool Bytes Max             | gauge   |
+| JVM Memory Pool Bytes Used            | gauge   |
+| JVM Threads Current                   | gauge   |
+| JVM Threads Daemon                    | gauge   |
+| JVM Threads Deadlocked                | gauge   |
+| JVM Threads Deadlocked Monitor        | gauge   |
+| JVM Threads Peak                      | gauge   |
+| JVM Threads Started Total             | gauge   |
+| Null                                  | counter |
```
Comment (jczuchnowski) [inline docs/supported-metrics.md:92]: Shouldn't it be `Forwarded Null` ?

File: docs/supported-metrics.md [hunk 11]
```diff
@@ -54,3 +54,48 @@ In Mesmer we support 3 types of metrics:
 | Stream throughput            | counter |
 | Operator throughput          | counter |
 | Operator processing time     | counter |
+
+## ZIO (both Default-JVM-Metrics and others)
```
Comment (jczuchnowski) [inline docs/supported-metrics.md:58]: General idea for this page - let's group the metrics in the following sections with different level headers:
```
# Supported metrics
## Akka
### (all current Akka sections)
## ZIO
### (all ZIO sections)
```
and in addition I'd split ZIO metrics into three sections like:
```
### Executor Metrics
### JVM Metrics
### ZIO Runtime Metrics

File: build.sbt (hunk 1 above)
Comment (ronaldSC-2022) [reply build.sbt]: @lgajowy, refactored in this commit:
https://github.com/ScalaConsultants/mesmer/pull/470/commits

I ended up going with the `get rid of "exampleDependencies" and just paste the dependencies here` suggestion 🙂

File: build.sbt (hunk 2 above)
Comment (ronaldSC-2022) [reply build.sbt]: @lgajowy, refactored in this commit:
https://github.com/ScalaConsultants/mesmer/pull/470/commits

File: docs/intro.md (hunk 5 above)
Comment (ronaldSC-2022) [reply docs/intro.md:17]: @lgajowy, refactored in this commit:
https://github.com/ScalaConsultants/mesmer/pull/470/commits

File: docs/supported-metrics.md (hunk 6 above)
Comment (ronaldSC-2022) [reply docs/supported-metrics.md]: @lgajowy, refactored in this commit:
https://github.com/ScalaConsultants/mesmer/pull/470/commits

File: docs/supported-metrics.md (hunk 8 above)
Comment (ronaldSC-2022) [reply docs/supported-metrics.md]: @lgajowy, refactored in this commit:
https://github.com/ScalaConsultants/mesmer/pull/470/commits

File: docs/supported-metrics.md (hunk 7 above)
Comment (ronaldSC-2022) [reply docs/supported-metrics.md]: @lgajowy, refactored in this commit:
https://github.com/ScalaConsultants/mesmer/pull/470/commits

Based off my testing (as seen in below screenshot), I could not find any `ZIO DefaultJvm _buckets metric in prometheus.` 🙁

File: docs/supported-metrics.md (hunk 10 above)
Comment (ronaldSC-2022) [reply docs/supported-metrics.md]: You are correct.

@jczuchnowski, I refactored the fix in this commit:
https://github.com/ScalaConsultants/mesmer/pull/470/commits/377f11eab460818b07249fe0b5c759f214d45413

File: docs/supported-metrics.md (hunk 11 above)
Comment (ronaldSC-2022) [reply docs/supported-metrics.md]: @jczuchnowski, I refactored the fix in this Commit:
https://github.com/ScalaConsultants/mesmer/pull/470/commits/377f11eab460818b07249fe0b5c759f214d45413

**Comment by ronaldSC-2022** [comment]
@jczuchnowski / @lgajowy, just a note:

For my changes in the `getting-started.MD` file, I got all of those information from Piotr Jósiak's Mesmer Evernote (see screenshot && URL below):

Piotr Mesmer EverNotes

**Mesmer Evernote URL:**
https://www.evernote.com/shard/s689/client/snv?noteGuid=c4321373-f2a7-706f-0b71-9a1dd72a0d03&noteKey=20bfb182ae3aa3fcd881328fb472c567&sn=https%3A%2F%2Fwww.evernote.com%2Fshard%2Fs689%2Fsh%2Fc4321373-f2a7-706f-0b71-9a1dd72a0d03%2F20bfb182ae3aa3fcd881328fb472c567&title=Links%2Bto%2Bget%2Bstarted

**Review by jczuchnowski — APPROVED** [review]

**Review by lgajowy — APPROVED** [review]

**Pull request merged**